opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py

```diff
@@ -286,15 +290,23 @@ def get_context(self):
     def set_attribute(
         self: "Span", key: str, value: types.AttributeValue
     ) -> None:
-        if self.attributes is Span.empty_attributes:
-            self.attributes = BoundedDict(MAX_NUM_ATTRIBUTES)
+        with self._lock:
+            if self.end_time is not None:
+                logger.warning("Calling set_attribute() on an ended span.")
```

Comment:
I am not 100% sure. If I move `if self.end_time is not None` outside the lock it could be racy with the `self.end_time = util.time_ns()` assignment in the `end()` method. Maybe I'm missing something...

Reply:
You must not move the condition outside the lock but you can do:

```python
with self.lock:
  has_ended = self.end_time is not None
if has_ended:
   # ...
```

Reply:
I refactored the code a little bit to avoid holding a lock while logging.